dask/dataframe/shuffle.py

```diff
@@ -24,6 +25,196 @@
 logger = logging.getLogger(__name__)
 
 
+class ShuffleStage(Layer):
+    """Shuffle HighLevelGraph Layer
+
+    High-level graph layer corresponding to a single stage of
+    an inter-partition shuffle operation.
+
+    Stage: (shuffle-group) -> (shuffle-split) -> (shuffle-join)
+
+    Parameters
+    ----------
+    name : str
+        Name of new (partially) shuffled collection.
+    column : str or list of str
+        Column(s) to be used to map rows to output partitions (by hashing).
+    inputs : list of tuples
+        Each tuple dictates the data movement for a specific partition.
+    stage : int
+        Index of the current shuffle stage.
+    npartitions : int
+        Number of output partitions for the full (multi-stage) shuffle.
+    n : int
+        Number of partitions in the original (un-shuffled) DataFrame.
+    k : int
+        A partition is split into this many groups during each stage.
+    ignore_index: bool, default False
+        Ignore index during shuffle.  If ``True``, performance may improve,
+        but index values will not be preserved.
+    dep_name : str
+        Name of input collection.
+    dep_meta : pd.DataFrame-like object
+        Empty metadata of input collection.
+    parts_out : list of int (optional)
+        List of required output-partition indices.
+    """
+
+    def __init__(
+        self,
+        name,
+        column,
+        inputs,
+        stage,
+        npartitions,
+        n,
+        k,
+        ignore_index,
+        dep_name,
+        dep_meta,
+        parts_out=None,
+    ):
+        self.column = column
+        self.name = name
+        self.inputs = inputs
+        self.stage = stage
+        self.npartitions = npartitions
+        self.n = n
+        self.k = k
+        self.ignore_index = ignore_index
+        self._name = dep_name
+        self._meta = dep_meta
+        self.parts_out = parts_out or range(len(inputs))
+
+    def __repr__(self):
+        return "ShuffleStage<name='{}', stage={}, k={}, npartitions={}>".format(
+            self.name, self.stage, self.k, self.npartitions
+        )
+
+    @property
+    def _dict(self):
+        """Materialize full dict representation"""
+        if hasattr(self, "_cached_dict"):
+            return self._cached_dict["dsk"]
+        else:
+            dsk = self._rbc_tasks_stage()
+            self._cached_dict = {"dsk": dsk}
+        return self._cached_dict["dsk"]
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __iter__(self):
+        return iter(self._dict)
+
+    def __len__(self):
+        return len(self._dict)
+
+    def _keys_to_parts(self, keys):
+        parts = set()
+        for key in keys:
+            try:
+                _name, _part = key
+            except ValueError:
+                continue
```

Comment:
I'm curious about when this occurs. We're saying that `keys` might be things like `(name, int)` or just something like `other_name`. Do you have a sense for which is more common? Catching exceptions can be relatively expensive if they happen often.

But perhaps this isn't the performance bottleneck and can be ignored.

Reply:
> I'm curious about when this occurs. We're saying that keys might be things like (name, int) or just something like other_name. Do you have a sense for which is more common? Catching exceptions can be relatively expensive if they happen often.

I don't believe this exception is very unlikely to happen (it never happens in any tests).  The purpose of this method is to extract a list of partition IDs (integers) from the keys being used to cull the `ShuffleStage` layer.  The input list is currently guaranteed to contain keys that look like: `(self.name, int)`.  Therefore, the `try/except` and `_name != self.name` logic is not expected to be used in practice (its only there to "harden" the method a bit)... Let me know if you'd like to be more explicit and just raise an error if the key is something unexpected.